cli: consul setup command

pkazmierczak opened the pull request:
An interactive setup helper for configuring Consul to accept Nomad WI-enabled workloads.

tgross commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+	"Description": "binding rule for Nomad workload identities (WI)",
+	"AuthMethod": "nomad-services",
+	"BindType": "service",
+	"BindName": "${value.nomad_namespace}-${value.nomad_service}"
```
After doing some end-to-end testing of reachability, it looks like we don't actually want this bit:

```suggestion
	"BindName": "${value.nomad_service}"
```

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+  # cert_file = "/etc/ssl/consul.crt"
+  # key_file  = "/etc/ssl/consul.key"
+
+  service_auth_method = "nomad-services"
+  task_auth_method    = "nomad-tasks"
```
```suggestion
```

I think we're already using the default values for these.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+		Type:          "jwt",
+		DisplayName:   authMethodName,
+		Description:   "login method for Nomad workload identities (WI)",
+		MaxTokenTTL:   defaultTTL,
```
```suggestion
```

I don't think we want a max TTL for Consul tokens since we can't renew them?

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+		Rules: string(policyBody),
+	}, nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "already exists") {
+			s.Ui.Warn(fmt.Sprintf("[ ] policy %s already exists", policyName))
+		}
```
Do we want to return early here like it's being done in the role function above?

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+	return nil
+}
+
+func (s *SetupConsulCommand) createRoleForTemplate() error {
```
```suggestion
func (s *SetupConsulCommand) createRoleForTasks() error {
```

I think?

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+func (s *SetupConsulCommand) askQuestion(question string) bool {
+	for {
+		answer, err := s.Ui.Ask(s.Colorize().Color(fmt.Sprintf("[?] %s", question)))
+		if err != nil {
+			if err.Error() != "interrupted" {
+				s.Ui.Output(err.Error())
+			}
+			return false
+		}
```
```suggestion
		if err != nil {
			if err.Error() != "interrupted" {
				s.Ui.Output(err.Error())
				os.Exit(1)
			}
			os.Exit(0)
		}
```

If we don't exit here it's impossible to `ctrl-C` the command since it will interpreted as a `no`.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+		}
+
+		switch strings.TrimSpace(strings.ToLower(answer)) {
+		case "", "n", "no":
+			return false
+		case "y", "yes":
+			return true
```
The questions have `[Y/n]` which usually indicate that `Y` is the default, but `case "", "n", "no"` causes `n` to be the default.

pkazmierczak replied:
ah, this is copy pasta from another command, my bad.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+		case "y", "yes":
+			return true
+		default:
+			s.Ui.Output(fmt.Sprintf("%s is not a valid response, please answer \"yes\" or \"no\".", answer))
```
```suggestion
			s.Ui.Output(fmt.Sprintf(`%q is not a valid response, please answer "yes" or "no".`, answer))
```

lgfa29 reviewed (changes requested):
Looking great! This will be super helpful for quick local testing. I just left some comments to improve the UX and some bugs I found.

pkazmierczak commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+var policyBody []byte
+
+const (
+	authMethodServices = "nomad-services"
+	authMethodTasks    = "nomad-tasks"
```
😬 yeah...

pkazmierczak commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+			s.Ui.Error(err.Error())
+			return 1
+		}
+	}
```
My reasoning was: if users reply `n` this indicates they either set some parts up themselves or prefer to set them up themselves. I think otherwise it makes no sense to have the interactive helper at all.

pkazmierczak commented on command/setup_consul.go:
```diff
@@ -0,0 +1,449 @@
+
+consul {
+  enabled = true
+  address = "<Consul address>"
+
+  # Nomad agents still need a Consul token in order to register themselves
+  # for automated clustering. It is recommended to set the token using the
+  # CONSUL_HTTP_TOKEN environment variable instead of writing it in the
+  # configuration file.
+
+  # Consul mTLS configuration.
+  # ssl       = true
+  # ca_file   = "/var/ssl/bundle/ca.bundle"
+  # cert_file = "/etc/ssl/consul.crt"
+  # key_file  = "/etc/ssl/consul.key"
```
I think it's ok to omit the mTLS config for this helper.

lgfa29 replied:
But if they use mTLS then this config will be incomplete and not work. Since we have access to the Consul client config we know exactly what needs to be set in the file.

Using a go template here could be useful, so something like this (totally untested 😅):

```
consul {
  enabled = true
  address = "{{.Address}}"

  # Nomad agents still need a Consul token in order to register themselves
  # for automated clustering. It is recommended to set the token using the
  # CONSUL_HTTP_TOKEN environment variable instead of writing it in the
  # configuration file.

  {{if .Namespace)}}
  namemespace = "{{.Namespace}}"
  {{end}}

  {{if (eq .Scheme "https")}}
  ssl       = true
  {{end}}
  {{if .TLSConfig.CAFile}}
  ca_file   = "{{.TLSConfig.CAFile}}"
  {{end}}
  {{if .TLSConfig.CertFile}}
  cert_file = "{{.TLSConfig.CertFile}}"
  {{end}}
  {{if .TLSConfig.CertFile}}
  key_file  = "{{.TLSConfig.KeyFile}}"
  {{end}}
  {{if .TLSConfig.InsecureSkipVerify}}
  verify_ssl  = "{{.TLSConfig.KeyFile}}"
  {{end}}
}
```

And then you can just feed it the Consul client config to get the correct Nomad config file.

Bonus point if you run the output through [`hclwrite.Format()`](https://pkg.go.dev/github.com/hashicorp/hcl2@v0.0.0-20191002203319-fb75b3253c80/hclwrite#Format) 😄

pkazmierczak commented:
@tgross @lgfa29 I addressed all the comments, mind giving it another look?

tgross commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	authMethodMsg := `
+We need to create two JWT auth methods: one for Nomad services, and one for
+Nomad tasks. The method for services will be called %v and the
+method for tasks %v, and they will both be of jwt type. They
+will share the following config:`
```
This is a minor fit-and-finish issue, but these lines aren't evenly wrapped.

pkazmierczak replied:
That's for a reason. I actually calculated the characters of method names that get inserted in this message to get perfect alignment. (I am not joking)

tgross replied:
Ah, perfect! Ship it! :grinning:

tgross commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+  enabled = true
+  address = "<Consul address>"
+
+  # Nomad agents still need a Consul token in order to register themselves
+  # for automated clustering. It is recommended to set the token using the
+  # CONSUL_HTTP_TOKEN environment variable instead of writing it in the
+  # configuration file.
```
:+1: I like how this text sets users up for a more secure approach.

tgross commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+
+	_, _, err := nsClient.Create(namespace, nil)
+	if err != nil {
+		return fmt.Errorf("[✘] could not write namespace %q: %w", consulNamespace, err)
```
Another minor fit-and-finish issue: we're inconsistent about using `%q` and `%s` for the user-supplied named. It might be nice if we stuck to one or the other throughout this command.

pkazmierczak replied:
good catch. Fixed in [6192f39](https://github.com/hashicorp/nomad/pull/18820/commits/6192f39c9ba160a73fbf37ae6b02f8475abb3026)

tgross reviewed (approved):
LGTM! I've left a few comments around specific bits of text but the overall workflow looks great.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	s.Ui.Output(`
+Finally, we need to create a role role-nomad-tasks associated with the policy
+above.`)
```
```suggestion
And finally, we will create an ACL role called %q associated with the policy
above.`, roleTasks)
```

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	var createRole bool
+	if !s.autoYes {
+		createRole = s.askQuestion(
+			"Should we create the above role in your Consul cluster? [Y/n]",
```
```suggestion
			"Create the above role in your Consul cluster? [Y/n]",
```

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	s.Ui.Output(`
+Finally, we need to create a role role-nomad-tasks associated with the policy
+above.`)
+
```
I think we're missing printing the role information.

pkazmierczak replied:
a role is just a role name we print above and a link to a policy.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+		Description: "binding rule for Nomad services w/ (WI)",
+		AuthMethod:  authMethodServices,
+		BindType:    "service",
+		BindName:    "${value.nomad_namespace}-${value.nomad_service}",
```
```suggestion
		BindName:    "${value.nomad_service}",
```

Explained in https://github.com/hashicorp/nomad/pull/18820#discussion_r1373773861.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	existingMethods, _, _ := s.client.ACL().AuthMethodList(nil)
+	if len(existingMethods) > 0 {
+		if slices.ContainsFunc(
```
```suggestion
		if slices.ContainsFunc(
```
I think we can omit this to make the code a bit easier to read? `ContainsFunc` seems to range over the slice, which will be a no-op on an empty slice anyway.

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	_, _, err := s.client.ACL().AuthMethodCreate(method, nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "error checking JWKSURL") {
+			s.Ui.Error("error: Nomad JWKS endpoint unreachable, is your Nomad server running and is the JWKS url set correctly?")
```
```suggestion
			s.Ui.Error("error: Nomad JWKS endpoint unreachable, verify that Nomad is running and that the JWKS URL %s is reachable by Consul", s.jwksURL)
```

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+		if slices.ContainsFunc(
+			existingMethods,
+			func(m *api.ACLAuthMethodListEntry) bool { return m.Name == method.Name }) {
+			s.Ui.Warn(fmt.Sprintf("[ ] auth method with name %s already exists", method.Name))
```
```suggestion
			s.Ui.Warn(fmt.Sprintf("[ ] auth method with name %q already exists", method.Name))
```

lgfa29 reviewed (commented):
Another round of me being annoying with details 😅

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	servicesBindingRule := &api.ACLBindingRule{
+		Description: "binding rule for Nomad services w/ (WI)",
```
```suggestion
		Description: "Binding rule for Nomad services authenticated using a workload identity",
```

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	tasksBindingRule := &api.ACLBindingRule{
+		Description: "binding rule for Nomad tasks w/ (WI)",
```
```suggestion
		Description: "Binding rule for Nomad tasks authenticated using a workload identity",
```

pkazmierczak commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	if createAuthMethods {
+		err = s.createAuthMethod(authMethodServices, authMethodConf)
+		if err != nil {
+			s.Ui.Error(err.Error())
+			return 1
+		}
+		err = s.createAuthMethod(authMethodTasks, authMethodConf)
+		if err != nil {
+			s.Ui.Error(err.Error())
```
`s.createAuthMethod` already returns an informative error.

pkazmierczak commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	if createBindingRules {
+		err = s.createBindingRules(servicesBindingRule)
+		if err != nil {
+			s.Ui.Error(err.Error())
+			return 1
+		}
+		err = s.createBindingRules(tasksBindingRule)
+		if err != nil {
+			s.Ui.Error(err.Error())
```
`s.createBindingRules` already returns an informative error.

pkazmierczak commented on command/setup_consul.go:
```diff
@@ -0,0 +1,485 @@
+	}
+
+	if s.consulEnt {
+		err = s.createNamespace()
```
We create it only when it doesn't exists, `s.createNamespace` checks that. We don't _need_ to create a namespace, but I thought this was considered good practice for consul enterprise to bind auth methods to a namespace?

tgross commented on command/setup_consul.go:
```diff
@@ -0,0 +1,483 @@
+// Help satisfies the cli.Command Help function.
+func (s *SetupConsulCommand) Help() string {
+	helpText := `
+Usage: nomad setup consul [options]
+
+  This command sets up Consul for allowing Nomad workloads to authenticate
+  themselves using Workload Identity.
+
+  This command requires acl:write permissions for Consul and respects
+  CONSUL_HTTP_TOKEN, CONSUL_HTTP_ADDR, and other Consul-related
+  environment variables as documented in
+  https://developer.hashicorp.com/nomad/docs/runtime/environment#summary. 
```
At one point we were discussing having `setup consul` and `setup vault` be specifically called out as experimental/non-stable interfaces, so that we could iterate on them more freely in the near future as users hit stumbling blocks during migration. Do we still want to do that here?

pkazmierczak replied:
good thinking, I added a warning in [5d12362](https://github.com/hashicorp/nomad/pull/18820/commits/5d12362b106b9d43415e0d949dad7b119fee0284)

pkazmierczak commented:
@lgfa29 I addressed all your comments and added a "cleanup" feature where we offer the possibility of removing all the configuration we created if users decide to stop ("atomicity").

lgfa29 commented on command/setup.go:
```diff
@@ -0,0 +1,39 @@
+	return strings.TrimSpace(helpText)
+}
+
+// Synopsis satisfies the cli.Command Synopsis function.
+func (a *SetupCommand) Synopsis() string { return "Interact with setup helpers" }
+
+// Name returns the name of this command.
+func (a *SetupCommand) Name() string { return "setup" }
+
+// Run satisfies the cli.Command Run function.
+func (a *SetupCommand) Run(_ []string) int { return cli.RunResultHelp }
```
```suggestion
// Synopsis satisfies the cli.Command Synopsis function.
func (c *SetupCommand) Synopsis() string { return "Interact with setup helpers" }

// Name returns the name of this command.
func (c *SetupCommand) Name() string { return "setup" }

// Run satisfies the cli.Command Run function.
func (c *SetupCommand) Run(_ []string) int { return cli.RunResultHelp }
```

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,736 @@
+	autoYes   bool
+
+	// store IDs of objects we may need to cleanup
+	policyID       string
```
```suggestion
```

I think we don't need this anymore. I avoid using the ID because this doesn't persist between runs, so the policy was not being cleaned up.

pkazmierczak replied:
fixed in [d04a235](https://github.com/hashicorp/nomad/pull/18820/commits/d04a235d90260a8dbac816af91512d5e5a20d93f)

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,736 @@
+
+	// store IDs of objects we may need to cleanup
+	policyID       string
+	bindingRuleIDs []string
```
Which makes me think about this one...I suspect we may be "leaking" the binding rules between runs 🤔 

Maybe we could use the `AuthMethod` value to find the ones we need to delete?

pkazmierczak replied:
fixed in [d04a235](https://github.com/hashicorp/nomad/pull/18820/commits/d04a235d90260a8dbac816af91512d5e5a20d93f)

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,736 @@
+to authenticate unless you create missing configuration yourself.
+`)
+
+	if !s.autoYes && s.askQuestion("Remove everything this command creates? [Y/n]") {
```
Can we list the things that will be deleted here? It could be a static list like:

```
[?] Remove everything this command creates? [Y/n]

The following items will be deleted if they exist:
  * Auth methods "nomad-services" and "nomad-tasks".
  * Binding rules associated with these methods.
  * Role named "role-nomad-task"
  * Policy named "nomad-workloads".
```

Or a dynamic list, where we first collect what needs to be deleted and output exactly what will be removed.

```
[?] Remove everything this command creates? [Y/n]

The following items will be deleted:
  * Auth methods "nomad-services" and "nomad-tasks".
  * Binding rules "ABC" and "XYZ".
  * Policy named "nomad-workloads.
```

pkazmierczak replied:
implemented in [599150e](https://github.com/hashicorp/nomad/pull/18820/commits/599150edb88dd997338d83c51c719014768a12fc)

lgfa29 commented on command/setup_consul.go:
```diff
@@ -0,0 +1,736 @@
+  -y
+    Automatically answers "yes" to all the questions, making the setup
+    non-interactive. Defaults to "false".
+
```
It could be useful to have a flag like `-cleanup` or `-remove` that just calls `handleNo()`. Otherwise it's impossible to remove things once the setup completes because it just run though automatically 😅

pkazmierczak replied:
implemented in [599150e](https://github.com/hashicorp/nomad/pull/18820/commits/599150edb88dd997338d83c51c719014768a12fc)

lgfa29 reviewed (approved):
Push some last minor changes, but this command is great! Thank you for the patience 😄
